geps/gep-1619.md

```diff
@@ -130,98 +275,563 @@ Generally, the implementation API programs the dataplane API; however these two
-| Nginx 	| Dataplane 	| [Consistent Hashing](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#hash) via IP or other connection attributes (Nginx Open Source) 	| N/A 	|
-| NGINX Kubernetes Gateway 	| Implementation (Nginx) 	| Supports Gateway API Only* 	| *NGINX Kubernetes Gateway solely uses Gateway API; therefore, it doesn’t yet have a way to configure session affinity. [Nginx Kubernetes Gateway Docs](https://github.com/nginxinc/nginx-kubernetes-gateway) 	|
-| Traefik 	| Implementation / Dataplane 	| None 	| [Traefik Docs](https://doc.traefik.io/traefik/) 	|
-
 ### Sessions in Java
 
-Java application servers such as Tomcat and Jetty, were the first to standardize the API around persistent sessions. These Java applications introduced the “jsessionid” cookie as well as more advanced features such as session migration, replication, and on demand session activation. It’s important for Gateway API to examine persistent session use cases and history from Java APIs to ensure the API is designed appropriately.
+Java application servers such as Tomcat and Jetty, were the first to standardize the API around cookies and sessions.
+These Java applications introduced the “jsessionid” cookie and session IDs encoded in URL parameters as well as more
+advanced features such as session migration, replication, and on demand session activation. It’s important for Gateway
+API to examine cookie use cases and history from Java APIs to ensure the API is designed appropriately.
 
-## API
-
-### Session Persistence API
-
-The majority of implementations and dataplanes support configuring cookies for session persistence while support for header-based session persistence is limited.
-
-To organize potential Gateway API configuration of cookie functionality, the next section will break down configuration into specific categories:
-- Session Cookie Configuration
-- Session Cookie Strategy
-
-**Session Cookie Configuration**
-
-Session cookie configuration are the parameters and attributes the dataplane uses to create a cookie for establishing a session.
+### Session Affinity in K8S Services
 
-A set-cookie request can be broken down into [3 main components](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie):
-- Cookie Name
-- Cookie Value
-- Cookie Attributes
+Kubernetes provides an API that allows you to enable [session affinity](https://kubernetes.io/docs/reference/networking/virtual-ips/#session-affinity)
+on service objects. It ensures consistent sessions by utilizing the client's IP address and also offers the option to
+set a timeout for the maximum session duration. Implementations of Gateway API don't rely on the service IP, so this
+particular application of session affinity doesn't impact the design of the GEP. However, we recommend that users enable
+session affinity using the Kubernetes service object to for applications requiring persistence, particularly due to the
+fact that in-cluster traffic will make use of the service IP.
```

Comment:
This might not be entirely correct, specifically in the GAMMA use case. 
Enabling K8s Service session affinity might also have negative effect as it puts more stringent restrictions on the LB decision vs just using application-level session semantics like cookie/headers.

Reply:
I still don't see any difference between the 2. In the early days of Servlet - client IP was considered a valid mechanism equivalent to cookies when used in local network ( so was TLS session ID ). There is nothing special about cookies/headers/url parameters - as long as the requirement is met. 

The stringent restrictions are on the application - it needs to be designed and configured to deal with the distorted balancing. 

Also: what other choices would a user have ? If they implement a service that requires sticky session today, they 
can either use Service affinity or stateful set. Internal traffic may not use the gateway, and our proposed solution is only for HTTP anyways while Service affinity works for TCP too.

Reply:
Agree with @va7ish. The idea is that some code in the application has the notion of a session and it indicates that by including a previously returned cookie.

Reply:
I agree as well - there are cases where users deploy a backend that require persistence and ONLY use it with HTTP gateways and cookies. That's very good and covered by this doc.

But there are also cases where users have such backends in use already - using the ClientIP affinity - or where the backends are accessed directly from cluster ( without a gateway or mesh). Right now in K8S this is supported via ClientIP affinity ( as well as TCP services which we don't cover with the cookies). 

What we need to address is how the 2 affinities interact - should we require that users who use the persistent backend only make calls trough Gateway ? Can the backend support both in-cluster IP affinity and cookie for gateway ? There are subtle failure modes - for example a gateway may be confused by the IP affinity and send all traffic to a single backend ( Istio doesn't implement/breaks client IP Service affinity and will be fine - but if it ever gets fixed - we'll be in trouble ).

So it's not about the clear cases where users only uses cookie ( and doesn't use Service v1 affinity ) - but I think we should at least say something for users who do use Service v1 affinity. Can be ' not supported, pick one'. And also if we decide to define 'affinity' and 'persistence' in a Gateaway - we should at least mention how 'v1 client ip affinity' relates to our definition. Are we re-defining the term, restricting it, replacing it - any choice is good, but ignoring it is IMO going to confuse users.

Reply:
The K8s Service clientIP affinity is a transport layer feature. Here we a discussing application layer session persistence. We could either say we support it which would mean that the more granular L7 affinity is entirely ignored (as the traffic is routed based on client's IP) or we ignore that K8s L4 configuration all together. I think the latter is the better statement as it would mean that existing traffic that flows outside of the Gateway/Mesh datapath is not broken (backwards compatible) while traffic that lands through Gateway/Mesh gets the functionality that was explicitly configured through the APIs in this GEP.